File: twoNumberSum.py
# Declare an initialize the main list variable
listOfNum = [2, 7, 8, 11, -1, -3]

# Find the two numbers in an array that satisfy a given target sum
# Complexity: O(n^2)) in time | O(1) in space
def twoNumberSum1(_list, targetSum):
    for i in range(0, len(_list)):
        firstNum = _list[i]
        for j in range(1, len(_list)):
            secondNum = _list[j]
            if firstNum + secondNum == targetSum:
                return [firstNum, secondNum]

    return []

# Find the two numbers in an array that satisfy a given target sum
# Complexity: O(n)) in time | O(n) in space
def twoNumberSum2(_list, targetSum):
    nums = {}
    for num in _list:
        potentialSum = targetSum - num
        if potentialSum in nums:
            return [potentialSum, num]
        else:
            nums[num] = True
    return []
# Find the two numbers in an array that satisfy a given target sum
# Complexity: O(nlog(n))) in time | O(1) in space
def twoNumberSum3(_list, targetSum):
    left = 0;
    right = len(_list) - 1
    _list.sort() # sort the list before making any calculate
    while left < right:
        if _list[left] + _list[right] == targetSum:
            return [_list[left], _list[right]]
        elif _list[left] + _list[right] < targetSum:
            left += 1
        elif _list[left] + _list[right] > targetSum:
            right -= 1

    return []

# Call of functions
print(twoNumberSum1(listOfNum, 10))
print(twoNumberSum2(listOfNum, 10))
print(twoNumberSum3(listOfNum, 10))

File: twoNumberSum.js
const num_array = [ -1, -4, 1, 3, 5, 6, 11, 8 ];
// Sort a given array by higher order
// Complexity: O(n) in time | O(1) in space
function sort_array(_array) {
	for (let i = 0; i < _array.length; i++) {
		var temp;
		if (_array[i] > _array[i + 1]) {
			temp = _array[i];
			_array[i] = _array[i + 1];
			_array[i + 1] = temp;
		}
	}

	return _array;
}

// Find the two numbers in an array that satisfy a given target sum
// Complexity: O(nlog(n)) in time | O(1) in space
function twoNumberSum1(sorted_array, targetSum) {
	var left = 0; // left pointer
	var right = sorted_array.length - 1; // right pointer
	var sum = 0;

	while (left < right) {
        sum = sorted_array[left] + sorted_array[right];
        if (sum === targetSum) {
            return [ sorted_array[left], sorted_array[right] ];
        } else if (sum > targetSum) {
			--right;
		} else if (sum < targetSum) {
			++left;
		}
	}

    return [null, null];
}

// Find the two numbers in an array that satisfy a given target sum
// Complexity: O(n^2)) in time | O(1) in space
function twoNumberSum2(_array, targetSum) {
    for (let i = 0; i < _array.length - 1; i++) {
        var firstNum = _array[i];
        for (let j = i + 1; j < _array.length; j++) {
            var secondNum = _array[j];

            if (firstNum + secondNum === targetSum) {
                return [firstNum, secondNum];
            }
        }
    }

    return [null, null];
}
// TODO: Find why the function don't return the correct Object
// Find the two numbers in an array that satisfy a given target sum
// Complexity: O(n) in time | O(n) in space
function twoNumberSum3(_array, targetSum) {

    var nums = {}; // Hash Table

    for (var i = 0; i < _array.length; i++) {
        var num = _array[i];
        var potentialMatchNum = targetSum - num;
        if (nums[potentialMatchNum] != undefined) {
            return [potentialMatchNum, num, nums];
        } else {
            nums[num] = true;
        }
    }

    return [null, null];
}
// call functions
console.log(sort_array(num_array));
//console.log(twoNumberSum2(sort_array(num_array), 6));
//console.log(twoNumberSum3(sort_array(num_array), -5));
